### PR TITLE
Add support of Duration field type #8875

### PR DESCRIPTION
**Pull request by denis**
It seems that it was already supported on the backend, so just added
support to the frontend. It only supports a single format now (h:mm:ss)
but it’s not very hard to add a dialog for allowing to choose format in
the future.

screen shot 2019-01-06 at 4 06 01 pm

This is my first PR so please let me know if I'm missing anything obvious.

###### Before submitting the PR, please make sure you do the following
- [x] If you're attempting to fix a translation issue, please submit your changes to our [POEditor project](https://poeditor.com/join/project/ynjQmwSsGh) instead of opening a PR.
### Tests
-  [x] Run the frontend and integration tests with  `yarn lint && yarn flow && yarn test`)
-  [x] If there are changes to the backend codebase, run the backend tests with `lein test && lein eastwood && lein bikeshed && lein docstring-checker && ./bin/reflection-linter`

-  [x] Sign the [Contributor License Agreement](https://docs.google.com/a/metabase.com/forms/d/1oV38o7b9ONFSwuzwmERRMi9SYrhYeOrkbmNaq9pOJ_E/viewform)
(unless it's a tiny documentation change).

### CODE REVIEW

**Comment by MarkRx** [comment]
A few notes:
* This assumes duration is in seconds but that is not universal. In our database we use millisecond precision. I don't think you can make an assumption on the granularity of the field. It may need to be configurable.
* Does this work if duration is null?
*

**Comment by denis** [comment]
@MarkRx both are valid, thanks for asking.

1. I now can see why in the original issue (#8875) it was referred as `Duration (seconds)` — to avoid confusion. I'm not sure what is the most common way to save durations, but I will stick with seconds (same way as UNIX timestamps are stored) but it will work for milliseconds as well if you use `Multiply by a number` formatting feature. As far as I know metabase doesn't offer any way to customize columns on data model level. I will check and make sure that `Multiply by a number` works as expected.

2. It works, since `null` (and `undefined`) cases are handled before the value goes into individual formatters.

File: frontend/src/metabase/lib/formatting.js [hunk 1]
```diff
@@ -470,6 +473,10 @@ export function formatTime(value: Value) {
   }
 }
 
+export function formatDuration(value: Value) {
+  return moment.duration(parseInt(value), "seconds").format("h:mm:ss");
```
Comment (MarkRx) [inline frontend/src/metabase/lib/formatting.js]: The problem with a hard coded format is you are assuming a duration range. In our case we have durations that last days, weeks, months, or even years.

**Comment by tlrobinson** [comment]
@salsakran @sbelak @camsaul If we want to support this should we add seconds/milliseconds subtypes types like we have for Unix timestamps?

```
(derive :type/DurationSeconds :type/Duration)
(derive :type/DurationMilliseconds :type/Duration)
```

**Comment by sbelak** [comment]
Can we use `:unit` instead like we do for bucketing?

**Comment by tlrobinson** [comment]
> Can we use `:unit` instead like we do for bucketing?

I don't think we currently persist `:unit` on Fields. It looks like it only appears on the pseudo-Fields we return when you start a question based on a Saved Question (presumably because they inherit from their columns' properties)

@camsaul @salsakran  Do we have a philosophy on what should go in the type hierarchy vs elsewhere?

e.x. we have the `currency` code in fields' `settings` object, but we also have things like Unix timestamp seconds/milliseconds in the type hierarchy.

I assume things that affect querying itself would typically be in the type hierarchy while purely display settings would be in `settings` (but maybe that won't always be the case)

**Review by mazameli — CHANGES_REQUESTED** [review]
I'd be down with adding this type. Per @tlrobinson's comments, I think it should be implemented similarly to `Currency`, where a second option is available to choose the unit for the duration. This would allow us to more easily extend this beyond `seconds` or `milliseconds` to whatever other units might make sense, like `hours`, `days` etc.

**Comment by tlrobinson** [comment]
This is still an open question I had:

> @camsaul @salsakran Do we have a philosophy on what should go in the type hierarchy vs elsewhere?
> 
> e.x. we have the currency code in fields' settings object, but we also have things like Unix timestamp seconds/milliseconds in the type hierarchy.
> 
> I assume things that affect querying itself would typically be in the type hierarchy while purely display settings would be in settings (but maybe that won't always be the case)

I'm not sure we need actually MBQL support for duration like we have for Unix timestamps. It will be a little weird to display durations like `hour:minute:second` in results, but in filters require you enter them as seconds or milliseconds depending on the underlying field type.

However, if we wanted to fix that, I think we could do it entirely on the frontend by showing pickers that transform to the correct unit (assuming 1 day = 24 h = 1440 minutes =  86400 seconds etc???), whereas I think for Unix timestamps the drivers need to be aware of the unit in order to handle extracts and truncating and timezones and all that fun stuff.

BTW there's a related issue with how we handle currency. Often currency is stored as cents which I don't believe we currently support. You can set the `Multiply by a number` setting on a field, but it's not taken into account when creating a filter.

@camsaul @sbelak @salsakran

**Comment by sbelak** [comment]
MBQL now has an `[:interval amount unit]` clause which is basically a duration.

**Comment by emmaferdez** [comment]
I started using Metabase for financial reports, and then I needed to report duration data. I've tried calculating it using SQL, but it's been a real headache. It would be great if you could at least add the small contributions you've made to this topic so we can start supporting durations or time intervals

**Pull request closed without merging**